Updated the clean URLs lesson to use dynamic routes (#365)
zeit/next-site: www/pages/learn/basics/clean-urls-with-dynamic-routing/dynamic-routing.mdx

import Layout from '../../../../components/learn/Layout';
import AnswerBox from '../../../../components/learn/AnswerBox';

export const meta = {
  title: 'Clean URLs with Route Masking',
  courseId: 'basics',
  lessonId: 'clean-urls-with-dynamic-routing',
  stepId: 'dynamic-routing',
  question: {
    answers: [
      'It throws an error.',
      'It goes back to the index page and navigates again to the post page.',
      'It goes back to the index page but nothing happens after that.',
      'It goes back to the index page then throws an error.'
    ],
    correctAnswer: 'It goes back to the index page and navigates again to the post page.'
  }
};

## Dynamic Routing

> Here, we are going to use a unique feature of Next.js called [Dynamic Routing](https://nextjs.org/docs#dynamic-routing), it allows you to create special pages in `/pages` to handle dynamic routes.

Reviewer: ```suggestion
Here, we are going to use the [Dynamic Routing](https://nextjs.org/docs#dynamic-routing) feature of Next.js, it allows you to handle dynamic routes in `/pages`.
```

Let's create our first dynamic route by adding a new page to `pages/p/[title].js`, notice that you need to add the folder `/p` inside `/pages` first, and that the page is called `[title].js`. Then add the following content:

```jsx
import { useRouter } from 'next/router';
import Layout from '../components/MyLayout.js';

export default function Post() {
  const router = useRouter();

  return (
    <Layout>
      <h1>{router.query.title}</h1>
      <p>This is the blog post content.</p>
    </Layout>
  );
}
```

> The previous page is special, instead of handling a static route like `/about`, it will handle any route that goes after `/p/`. For example `/p/hello-nextjs` will be handled by this page.

Reviewer: ```suggestion
The previous page is special, instead of handling a static route like `/about`, it will handle routes that come after `/p/`. For example, `/p/hello-nextjs` will be handled by this page. Although, `/p/p/another` will not.
```

> The page is also surrounded by brackets (`[]`), that's the wildcard, meaning that if you add brackets to the file name of a page it will then be a dynamic route.

Reviewer: ```suggestion
Having brackets (`[]`) in the page name makes it a dynamic route. Currently, you can not make part of a page name dynamic only the full name. `/pages/posts/[post].js` is supported but `/pages/posts/post-[id].js` is not currently. 
```

> And what's `title` for? It will be the name of the query param received by the page, so for `/p/hello-nextjs` the `query` object will have `{ title: 'hello-nextjs'}`, and we access to it with [useRouter()](https://nextjs.org/docs#userouter).

Reviewer: ```suggestion
When creating the dynamic route we added `title` between the brackets (`[]`). This is the name of the query param received by the page, so for `/p/hello-nextjs` the `query` object will have `{ title: 'hello-nextjs'}`, and we access to it with [useRouter()](https://nextjs.org/docs#userouter).
```

---

Now let's add multiple links that use our new dynamic route, Use the following code for `pages/index.js`:

```jsx
import Layout from '../components/MyLayout.js';
import Link from 'next/link';

const PostLink = props => (
  <li>

>     <Link href="/p/[title]" as={`/p/${props.id}`}>

Reviewer: This is confusing, why do `props.id` here but call the dynamic route `title`? Why not `/posts/[id]`?

      <a>{props.title}</a>
    </Link>
  </li>
);

export default function Blog() {
  return (
    <Layout>
      <h1>My Blog</h1>
      <ul>
        <PostLink id="hello-nextjs" />
        <PostLink id="learn-nextjs" />
        <PostLink id="deploy-nextjs" />
      </ul>
    </Layout>
  );
}
```

Have a look at the following code block:

```jsx
const PostLink = props => (
  <li>
    <Link href="/p/[title]" as={`/p/${props.id}`}>
      <a>{props.title}</a>
    </Link>
  </li>
);
```

> In the `<Link>` element, the `href` prop is now the path of the page in the `pages` folder. We also added a new prop named `as`, that's the URL we need to show on the browser.

Reviewer: ```suggestion
In the `<Link>` element, the `href` prop is now the path of the page in the `pages` folder and `as` is the URL to show in URL bar of the browser.
```

`as` isn't a new prop
